cmd/collect-profiles/main.go

```diff
@@ -0,0 +1,364 @@
+	rootCmd.MarkFlagRequired("config-mount-path")
+	rootCmd.PersistentFlags().StringVarP(&certMountPath, "cert-mount-path", "", "/var/run/secrets/serving-cert", "The path to the tls cert used by the client making https requests against the pprof URLs.")
+}
+
+func main() {
+	Execute()
+}
+
+func Execute() {
+	if err := rootCmd.Execute(); err != nil {
+		klog.Fatal(err)
+		os.Exit(1)
+	}
+}
+
+func getTruePointer() *bool {
+	trueBool := true
+	return &trueBool
+}
+
+func newCmd() *cobra.Command {
+	var cfg config.Configuration
+	return &cobra.Command{
+		Use:   "collect-profiles configMapName:url",
+		Short: "Retrieves the pprof data from a UNL and stores it in a configMap.",
+		Long: `The collect-profiles command makes https requests against pprof URLs
+		provided as arguments and stores that information in immutable configMaps.`,
+		Version: "0.0.1",
+		PersistentPreRunE: func(*cobra.Command, []string) error {
+			return cfg.Load()
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+
+			if len(args) == 0 {
+				klog.Info("No arguments provided, exiting")
+				return nil
+			}
+
+			jobConfig, err := config.GetConfig(configMountPath)
+			if err != nil {
+				klog.Infof("error retrieving job config")
+				return err
+			}
+
+			// Exit if job is disabled
+			if jobConfig.Disabled {
+				klog.Infof("CronJob disabled, exiting")
+				return nil
+			}
+
+			// Validate input
+			validatedArguments := make([]*argument, len(args))
+			for i, arg := range args {
+				a, err := newArgument(arg)
+				if err != nil {
+					return err
+				}
+				validatedArguments[i] = a
+			}
+
+			// Get existing configmaps
+			existingConfigMaps := &corev1.ConfigMapList{}
+			if err := cfg.Client.List(cmd.Context(), existingConfigMaps, client.InNamespace(namespace), client.HasLabels{profileConfigMapLabelKey}); err != nil {
+				return err
+			}
+
+			newestConfigMaps, expiredConfigMaps := separateConfigMapsIntoNewestAndExpired(existingConfigMaps.Items)
+
+			// Attempt to delete all but the newest configMaps generated by this job
+			errs := []error{}
+			for _, cm := range expiredConfigMaps {
+				if err := cfg.Client.Delete(cmd.Context(), &cm); err != nil {
+					errs = append(errs, err)
+					continue
+				}
+				klog.Infof("Successfully deleted configMap %s/%s", cm.GetNamespace(), cm.GetName())
+			}
+
+			// If a delete call failed, abort to avoid creating new configMaps
+			if len(errs) != 0 {
+				return fmt.Errorf("error deleting expired pprof configMaps: %v", errs)
+			}
+
+			certPath := filepath.Join(certMountPath, corev1.TLSCertKey)
+			keyPath := filepath.Join(certMountPath, corev1.TLSPrivateKeyKey)
+
+			if err := verifyCertAndKeyExist(certPath, keyPath); err != nil {
+				klog.Infof("error verifying provided cert and key: %v", err)
+				klog.Info("generating a new cert and key")
+				return populateServingCert(cmd.Context(), cfg.Client)
```

Comment:
Should we be returning here, or just attempting to create/update/etc. this serving cert secret? What happens if this cert/key path doesn't exist? Do we run this binary once, those files are created, and we run this binary again?

Reply:
>Should we be returning here, or just attempting to create/update/etc. 

We should return here. The CronJob runs every 15 minutes. Mounted secrets are only refreshed every 15 minutes. Even if we update and create these files, the OLM and Catalog pods will take at least a minute to notice and respond to the change.

Reply:
>What happens if this cert/key path doesn't exist?

If the secret doesn't exist an error is returned.
If the files are empty (which happens on the first run) this method is called earlier and initializes the values.

>Do we run this binary once, those files are created, and we run this binary again?

Correct. First run is just to initialize the secret. This happens once on a cluster. Subsequent runs (even on upgrades) do not require initialization.